File: level3/정수삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int t_size = triangle.size(), answer = 0;
    vector<vector<int>> dp(t_size);
    dp[0].push_back(triangle[0][0]);
    for(int i = 1; i < t_size; i++){
        for(int j = 0; j <= i; j++){
            if(j == 0){
                dp[i].push_back(triangle[i][j] + dp[i-1][j]);
            }
            else if(j == i){
                dp[i].push_back(triangle[i][j] + dp[i-1][j-1]);
            }
            else{
                dp[i].push_back(triangle[i][j] + max(dp[i-1][j-1],dp[i-1][j]));
            }
            answer = max(answer,dp[i][j]);
        }
    }
    return answer;
}

File: level3/정수삼각형.py
def solution(triangle):
    t_sum = [[0]*(i+2) for i in range(1,len(triangle)+1)]
    t_sum[0][1] = triangle[0][0]
    for i in range(1,len(t_sum)):
        for j in range(1,len(t_sum[i])-1):
            t_sum[i][j] = max(t_sum[i-1][j-1], t_sum[i-1][j]) + triangle[i][j-1]
    print(t_sum)
    return max(t_sum[-1])
